lib/spack/spack/environment.py

```diff
@@ -511,47 +564,78 @@ def add(self, user_spec):
+
+        added = False
+        existing = False
+        for i, (name, speclist) in enumerate(self.read_specs.items()):
+            if name == list_name:
+                # TODO: Add conditional which reimplements name-level checking
+                existing = str(spec) in speclist.yaml_list
+                if not existing:
+                    speclist.add(str(spec))
+                    added = True
+            elif added:
+                new_reference = dict((n, self.read_specs[n])
+                                     for n in list(self.read_specs.keys())[:i])
+                speclist.update_reference(new_reference)
         return bool(not existing)
 
-    def remove(self, query_spec, force=False):
+    def remove(self, query_spec, list_name='specs', force=False):
         """Remove specs from an environment that match a query_spec"""
         query_spec = Spec(query_spec)
 
-        # try abstract specs first
-        matches = []
-        if not query_spec.concrete:
-            matches = [s for s in self.user_specs if s.satisfies(query_spec)]
-
-        if not matches:
-            # concrete specs match against concrete specs in the env
-            specs_hashes = zip(
-                self.concretized_user_specs, self.concretized_order)
-            matches = [
-                s for s, h in specs_hashes if query_spec.dag_hash() == h]
-
-        if not matches:
-            raise SpackEnvironmentError("Not found: {0}".format(query_spec))
-
-        for spec in matches:
-            if spec in self.user_specs:
-                self.user_specs.remove(spec)
-
-            if force and spec in self.concretized_user_specs:
-                i = self.concretized_user_specs.index(spec)
-                del self.concretized_user_specs[i]
-
-                dag_hash = self.concretized_order[i]
-                del self.concretized_order[i]
-                del self.specs_by_hash[dag_hash]
+        removed = False
+        for i, (name, speclist) in enumerate(self.read_specs.items()):
+            if name == list_name:
+                # try abstract specs first
+                matches = []
+
+                if not query_spec.concrete:
+                    matches = [s for s in speclist if s.satisfies(query_spec)]
+
+                if not matches:
+                    # concrete specs match against concrete specs in the env
+                    specs_hashes = zip(
+                        self.concretized_user_specs, self.concretized_order)
+                    matches = [
+                        s for s, h in specs_hashes
+                        if query_spec.dag_hash() == h
+                    ]
+
+                if not matches:
+                    raise SpackEnvironmentError(
+                        "Not found: {0}".format(query_spec))
+
+                for spec in matches:
+                    if spec in speclist:
+                        speclist.remove(spec)
```

Comment:
If I understand correctly `SpecList` is responsible for expanding matrices to lists of user specs. Will the updated `spack.yaml` expand any matrix to a list of user specs? If not, how is removing specs from such `SpecList`s handled?

Reply:
A user can only remove a spec that is a direct entry in the given list. There isn't (at least currently) any support for removing a spec from within a matrix, because I'm not confident that the expected behavior is even well-defined in that case. I believe I address this in the documentation.

If you want to remove from the matrix lists, use a named list in the matrix and remove it from the named list.

Reply:
> If you want to remove from the matrix lists, use a named list in the matrix and remove it from the named list.

If that is possible now then the current logic may be problematic: multiple specs in `concretized_specs` may match the spec to remove, so the `.index` logic below would be sufficient.

Reply:
You can't remove from a matrix.

If the matrix refers to another list, you can remove from that list to affect the matrix.

Reply:
Perhaps I misunderstand the meaning of a list reference in a matrix: I assumed that if a list contains `[a, b]` and you had a matrix like:

```
matrix:
  - $list
  - [x, y, z]
```

That you would end up with 6 specs installed. Removing 1 item from the list would remove 3 items from the expanded matrix. If that is correct, then I think my above concern applies. Otherwise I'm probably off base.

Reply:
Yes, removing something from `list` would remove three things from the matrix. But that's exactly what you want in some cases.

```
matrix:
  - $compilers
  - $packages
```

Then when I remove `'%gcc@4.9.3` from `compilers`, what I want is the end result that every `foo%gcc@4.9.3` for `foo` in `$packages` is removed from the end result.

Reply:
Oh I just realized what you're talking about! The issue is how we handle the `force` argument when we're removing multiple specs (or any time we're not removing from the list named `specs`)... I'll have to think through this and see whether it will work.

Reply:
You're correct that this would be an issue. I've written a test that fails for it as-is and updated the code to account for it.